### DISCUSSION TITLE
Questions on hooks execution and queuing

### DISCUSSION THREAD

**Discussion by adminorcail** [pocketbase/pocketbase#5912] [Q&A] [answered]
Hi ! First of all, thanks for this project, that's awesome !

I have questions related to hooks:
- are pending/executing hooks completed before Pocketbase is shutted down ? (is the binary replacement "forced" ?)
- what happens if Pocketbase crashed or the binary is replaced while hooks are "pending" ? (are hooks queued somehow ?)

Here's why : to deploy Pocketbase (on Fly.io or elsewhere), one can just replace the Pocketbase binary ([from your guide on how to deploy on Fly.io](https://github.com/pocketbase/pocketbase/discussions/537)). Let's say that I deploy at a time where there are requests coming to Pocketbase, and that hooks should be executed (or are pending) at that time (for example, a record has been created and a onAfterCreated hook is executing, but has not finished yet). What will happened to these hooks ? Won't they be canceled because of the binary replacement ? In that case, is it really safe to deploy that way ? From my current understanding, the Pocketbase instance is not stopped before being replaced, so its execution won't be stopped properly and the hooks won't be resumed when Pocketbase will be restarted again.

Apart from deployment, my related use case is : I have transactions that create dependent/related records (if a record from collection is created, other records from other collections should be created as well, so I just use transactions to do so). This works well. But I wonder if I could replace that transaction with cascading hooks, as they are much more handy to work with (code-side).

I don't have the need to handle critical hooks execution (I know how to implements such architecture and tools to do so), I just wonder how to use and maintain Pocketbase as best as possible (and how to made my life simpler).

**Comment by ganigeorgiev** [comment, marked as answer]
> are pending/executing hooks completed before Pocketbase is shutted down ?

If the application process was gracefully interrupted (e.g. received `SIGTERM`) then it would wait couple seconds for the remaining requests to complete before exiting.

> what happens if Pocketbase crashed

If the process is abruptly terminated or killed by the OS (e.g. because of OOM) then nothing happens, aka. there is no queue for the execution out of the box to somehow "resume" your previous app process operations and I'm not sure if this can be generalized because it depends too much on the task(s) that need to be resumed (_e.g. what if it relies on the request context state or to send http response to a connection that no longer will be available?_).

>  binary is replaced while hooks are "pending"

At least on Linux, it is safe to remove/replace a running executable because the related inode won't be released until the process exit (and the graceful shutdown handling would take its turn).

For fly.io specifically you can check https://github.com/pocketbase/pocketbase/discussions/4706.

> But I wonder if I could replace that transaction with cascading hooks, as they are much more handy to work with (code-side).

Your description is too vague to understand what do you mean but note that with the upcoming v0.23.0 release users can wrap the entire hooks chain in a transaction if they really need to.

**Reply by adminorcail** [reply]
Thanks for your response !

> Your description is too vague to understand what do you mean but note that with the upcoming v0.23.0 release users can wrap the entire hooks chain in a transaction if they really need to.

Yes, this is because this is a more broad question than what my use case covers. I will look forward this release, as I need that for some hooks chain.

Thanks again.
